pyat/at/lattice/lattice_object.py

```diff
@@ -6,104 +6,154 @@
 
 TWO_PI_ERROR = 1.E-4
 
 
+def _scan_list(descr, keep_all=False, **kwargs):
+    """Extract the lattice attributes from an element list"""
+    attrs = {'name': '', '_radiation_on': False}
+    params = [elem for elem in descr if isinstance(elem, elements.RingParam)]
+    if params:
+        # Set all RingParam attributes to the lattice object
+        attrs.update(
+            (Lattice._translate.get(key, key.lower()), value)
+            for (key, value) in vars(params[0]).items()
+            if key not in Lattice._ignore)
+
+        if not keep_all:
+            descr = [elem for elem in descr
+                     if not isinstance(elem, elements.RingParam)]
+
+    for elem in descr:
+        if elem.PassMethod.endswith('RadPass') or elem.PassMethod.endswith(
+                'CavityPass'):
+            attrs['_radiation_on'] = True
+            break
+
+    attrs.update(kwargs)
+
+    if attrs.get('energy') is None:
+        # Guess energy from the Energy attribute of the elements
+        # Look first in cavities
+        energies = [elem.Energy for elem in descr
+                    if isinstance(elem, elements.RFCavity)]
+        if not energies:
+            # Then look in all elements
+            energies = [elem.Energy for elem in descr
+                        if hasattr(elem, 'Energy')]
+        if not energies:
+            raise AtError('Lattice Energy is not defined')
+        energy = max(energies)
+        if min(energies) < energy:
+            warn(AtWarning(
+                'Inconsistent energy values, "Energy" set to {0}'.format(
+                    energy)))
+        attrs['energy'] = energy
+
+    if attrs.get('periodicity') is None:
+        # Guess periodicity from the bending angle of the superperiod
+        theta = [elem.BendingAngle for elem in descr if
+                 isinstance(elem, elements.Dipole)]
+        try:
+            nbp = 2.0 * pi / sum(theta)
+        except ZeroDivisionError:
+            warn(AtWarning('No bending in the cell, set "Periodicity" to 1'))
+            attrs['periodicity'] = 1
+        else:
+            periodicity = int(round(nbp))
+            if abs(periodicity - nbp) > TWO_PI_ERROR:
+                warn(AtWarning(
+                    'Non-integer number of cells: {0} -> {1}'.format(
+                        nbp, periodicity)))
+            attrs['periodicity'] = periodicity
+
+    return descr, attrs
+
+
 class Lattice(list):
     """Lattice object
-    An AT lattice is a sequence of AT elements. This object accepts extended indexing"""
+    An AT lattice is a sequence of AT elements.
+    This object accepts extended indexing"""
 
-    _translate = dict(Energy='energy', Periodicity='periodicity', FamName='name')
+    _translate = {'Energy': 'energy', 'Periodicity': 'periodicity',
+                  'FamName': 'name'}
     _ignore = {'PassMethod', 'Length'}
 
-    def __init__(self, descr=None, **kwargs):
+    def __init__(self, descr=None, _scan=True, **kwargs):
```

Comment:
Could we give `descr` a more informative name please, maybe `elements`? It would also be nice to list more of the arguments here i.e. `def __init__(self, elements=None, name='lattice', energy=None, periodicity=None, keep_all=False, **kwargs):`

Reply:
You are right, `elements` looks more adequate.